tracer/src/Datadog.Trace/DataStreamsMonitoring/PathwayContextEncoder.cs

```diff
@@ -0,0 +1,77 @@
+
+    /// <summary>
+    /// Encodes a <see cref="PathwayContext"/> as a series of bytes
+    /// NOTE: the encoding is lossy, in that we convert <see cref="PathwayContext.PathwayStart"/>
+    /// and <see cref="PathwayContext.EdgeStart"/> from ns to ms
+    /// </summary>
+    /// <param name="pathway">The pathway to encode</param>
+    /// <returns>The encoded pathway</returns>
+    public static byte[] Encode(PathwayContext pathway)
+    {
+        var pathwayStartMs = pathway.PathwayStart / 1000;
+        var edgeStartMs = pathway.EdgeStart / 1000;
+
+        var pathwayBytes = VarEncodingHelper.VarLongZigZagLength(pathwayStartMs);
+        var edgeBytes = VarEncodingHelper.VarLongZigZagLength(edgeStartMs);
+
+        // maximum size = 8 + edge + pathway
+        var bytes = new byte[8 + pathwayBytes + edgeBytes];
+        BinaryPrimitivesHelper.WriteUInt64LittleEndian(bytes, pathway.Hash.Value);
+        VarEncodingHelper.WriteVarLongZigZag(bytes, offset: 8, pathwayStartMs);
+        VarEncodingHelper.WriteVarLongZigZag(bytes, offset: 8 + pathwayBytes, edgeStartMs);
+        return bytes;
+    }
+
+    /// <summary>
+    /// Tries to decode a <see cref="PathwayContext"/> from a <c>byte[]</c>.
+    /// NOTE: the encoding process is lossy, so the decoded <see cref="PathwayContext"/>
+    /// contains truncated values for <see cref="PathwayContext.PathwayStart"/>
+    /// and <see cref="PathwayContext.EdgeStart"/> (conversion from ns to ms)
+    /// </summary>
+    /// <param name="bytes">The pathway to decode</param>
+    /// <returns>The decoded <see cref="PathwayContext"/>, or <c>null</c> if decoding fails </returns>
+    public static PathwayContext? Decode(byte[] bytes)
+    {
+        if (bytes.Length < 10)
+        {
+            Log.Warning<string, int>("Error decoding Data Stream PathwayContext from bytes {Base64EncodedBytes}: insufficient bytes ({ByteCount})", Convert.ToBase64String(bytes), bytes.Length);
+            return null;
+        }
+
+        // first 8 bytes
+        var hash = BinaryPrimitivesHelper.ReadUInt64LittleEndian(bytes);
+
+        var pathwayStartMs = VarEncodingHelper.ReadVarLongZigZag(bytes, offset: 8, out var bytesRead);
+        if (pathwayStartMs is null)
+        {
+            Log.Warning("Error decoding Data Stream PathwayContext from bytes {Base64EncodedBytes}: invalid pathway start", Convert.ToBase64String(bytes));
+            return null;
+        }
+
+        var edgeBytesMs = VarEncodingHelper.ReadVarLongZigZag(bytes, offset: 8 + bytesRead, out _);
+        if (edgeBytesMs is null)
+        {
+            Log.Warning("Error decoding Data Stream PathwayContext from bytes {Base64EncodedBytes}: invalid edge start", Convert.ToBase64String(bytes));
+            return null;
+        }
+
+        // Pathway context values are in ns
+        return new PathwayContext(new PathwayHash(hash), pathwayStartMs.Value * 1000, edgeBytesMs.Value * 1000);
```

Comment:
Should the `pathwayStartMs.Value * 1000` be `pathwayStartMs.Value * 1000 * 1000` to get to nanoseconds from milliseconds?

Reply:
Oh wait, maybe I'm misreading `Ms` as milliseconds and it is actually microseconds?

As the conversions above (lines 27/28) got it as `var pathwayStartMs = pathway.PathwayStart / 1000;` which appears to be microseconds

So maybe if they are microseconds swap the `Ms` to `Us`?
Or it makes for longer names, but I tend to prefer to spell out the name of the units entirely, but that is more of a personal preference

Reply:
You're absolutely right, it's supposed to be milliseconds, so fixed the conversion 🤦 Thanks!